<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started with VDJigsaw • VDJigsaw</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with VDJigsaw">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">VDJigsaw</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/VDJigsaw_Vignette_01.html">Getting Started with VDJigsaw</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/doRemy/VDJigsaw"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Getting Started with VDJigsaw</h1>
                        <h4 data-toc-skip class="author">Rémy
Pétremand</h4>
            
            <h4 data-toc-skip class="date">16 February, 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/doRemy/VDJigsaw/blob/main/vignettes/VDJigsaw_Vignette_01.Rmd" class="external-link"><code>vignettes/VDJigsaw_Vignette_01.Rmd</code></a></small>
      <div class="d-none name"><code>VDJigsaw_Vignette_01.Rmd</code></div>
    </div>

    
    
<style type="text/css">
pre {
  max-height: 600px;
  overflow-y: auto;
}

pre[class] {
  max-height: 300px;
}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette walks you through a typical VDJigsaw workflow using
real-world VDJ data from 10x Genomics. By the end, you will have
assigned clonotypes at multiple stringency levels and visualized the
results.</p>
<p><strong>What does VDJigsaw do?</strong> Single-cell TCR sequencing
often suffers from dropout: a cell’s second alpha or beta chain allele
may simply not get captured. Most tools treat missing chains as
“different”, fragmenting clonotypes. VDJigsaw instead looks across the
whole dataset to puzzle together which cells likely belong to the same
clone, even when some chain data is missing. It does this at five levels
of stringency, from strict (all four chains must match) to permissive (a
single allele suffices).</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://doRemy.github.io/VDJigsaw/">VDJigsaw</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>To illustrate VDJigsaw, we use VDJ filtered contig data from two
publicly available 10x Genomics datasets:</p>
<ol style="list-style-type: decimal">
<li><p><strong>10k Human Diseased PBMCs (ALL) — Freshly
Processed</strong> (<a href="https://www.10xgenomics.com/datasets/10k_5p_Human_diseased_PBMC_ALL_Fresh" class="external-link">dataset</a>
| <a href="https://cf.10xgenomics.com/samples/cell-vdj/8.0.1/10k_5p_Human_diseased_PBMC_ALL_Fresh/10k_5p_Human_diseased_PBMC_ALL_Fresh_vdj_t_filtered_contig_annotations.csv" class="external-link">VDJ-TCR
contigs</a>)</p></li>
<li><p><strong>10k Human Diseased PBMCs (ALL) — Fixed and
Stored</strong> (<a href="https://www.10xgenomics.com/datasets/10k_5p_Human_diseased_PBMC_ALL_Fix_stored" class="external-link">dataset</a>
| <a href="https://cf.10xgenomics.com/samples/cell-vdj/8.0.1/10k_5p_Human_diseased_PBMC_ALL_Fix_stored/10k_5p_Human_diseased_PBMC_ALL_Fix_stored_vdj_t_filtered_contig_annotations.csv" class="external-link">VDJ-TCR
contigs</a>)</p></li>
</ol>
<p>These CSV files are bundled with the package in
<code>inst/extdata/</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VDJ.contigs.fresh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>,</span>
<span>  <span class="st">"10k_5p_Human_diseased_PBMC_ALL_Fresh_vdj_t_filtered_contig_annotations.csv"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"VDJigsaw"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">VDJ.contigs.fix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>,</span>
<span>  <span class="st">"10k_5p_Human_diseased_PBMC_ALL_Fix_stored_vdj_t_filtered_contig_annotations.csv"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"VDJigsaw"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For simplicity, we tag each dataset with a sample label and combine
them:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VDJ.contigs.fresh</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">&lt;-</span> <span class="st">"Fresh"</span></span>
<span><span class="va">VDJ.contigs.fix</span><span class="op">$</span><span class="va">orig.ident</span>   <span class="op">&lt;-</span> <span class="st">"Fix"</span></span>
<span><span class="va">VDJ.contigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">VDJ.contigs.fresh</span>, <span class="va">VDJ.contigs.fix</span><span class="op">)</span></span></code></pre></div>
<p>Let’s take a quick look at the raw data — each row is one contig (one
chain from one cell):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">VDJ.contigs</span><span class="op">)</span>, <span class="st">"contigs across"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">VDJ.contigs</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span><span class="op">)</span>, <span class="st">"unique barcodes\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 27129 contigs across 13634 unique barcodes</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">VDJ.contigs</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"chain"</span>, <span class="st">"v_gene"</span>, <span class="st">"cdr3"</span>, <span class="st">"j_gene"</span>, <span class="st">"umis"</span>, <span class="st">"orig.ident"</span><span class="op">)</span><span class="op">]</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="24%">
<col width="7%">
<col width="11%">
<col width="24%">
<col width="10%">
<col width="6%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th align="left">barcode</th>
<th align="left">chain</th>
<th align="left">v_gene</th>
<th align="left">cdr3</th>
<th align="left">j_gene</th>
<th align="right">umis</th>
<th align="left">orig.ident</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AAACCAAAGAACAGAC-1</td>
<td align="left">TRB</td>
<td align="left">TRBV20-1</td>
<td align="left">CSARDLYRGRETQYF</td>
<td align="left">TRBJ2-5</td>
<td align="right">25</td>
<td align="left">Fresh</td>
</tr>
<tr class="even">
<td align="left">AAACCAAAGAACAGAC-1</td>
<td align="left">TRA</td>
<td align="left">TRAV8-4</td>
<td align="left">CAVSDRQYSGGGADGLTF</td>
<td align="left">TRAJ45</td>
<td align="right">10</td>
<td align="left">Fresh</td>
</tr>
<tr class="odd">
<td align="left">AAACCAAAGAACAGAC-1</td>
<td align="left">TRA</td>
<td align="left">TRAV19</td>
<td align="left">CALSGQTSGTYKYIF</td>
<td align="left">TRAJ40</td>
<td align="right">11</td>
<td align="left">Fresh</td>
</tr>
<tr class="even">
<td align="left">AAACCAAAGCAAGATA-1</td>
<td align="left">TRA</td>
<td align="left">TRAV30</td>
<td align="left">CGTYNSGNTGKLIF</td>
<td align="left">TRAJ37</td>
<td align="right">10</td>
<td align="left">Fresh</td>
</tr>
<tr class="odd">
<td align="left">AAACCAAAGCAAGATA-1</td>
<td align="left">TRB</td>
<td align="left">TRBV11-2</td>
<td align="left">CASSLGRGRVAYEQFF</td>
<td align="left">TRBJ2-1</td>
<td align="right">7</td>
<td align="left">Fresh</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="run-the-pipeline">Run the pipeline<a class="anchor" aria-label="anchor" href="#run-the-pipeline"></a>
</h2>
<p>The main function <code><a href="../reference/assign_clonotype.html">assign_clonotype()</a></code> runs the full
VDJigsaw pipeline in four steps:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Pivot</strong> the long-format contigs into wide format (one
row per cell, with TRA_1, TRA_2, TRB_1, TRB_2 columns)</li>
<li>
<strong>Validate</strong> TCR chains (fix ordering, remove
duplicates, flag invalid sequences)</li>
<li>
<strong>Identify clonotypes</strong> at five stringency levels using
graph-based clustering</li>
<li>
<strong>Return</strong> annotated data and reference tables</li>
</ol>
<p>You can set <code>sample_col</code> to tell VDJigsaw about your
sample structure. The <code>clone_loose</code> parameter controls which
stringency level becomes the default <code>CloneID</code>. Set
<code>num_cores</code> to speed up processing when you have multiple
samples (use <code>-1</code> for all available cores).</p>
<blockquote>
<p><strong>Tip:</strong> Ideally, run VDJigsaw after quality control of
your scRNA-seq data to remove low-quality barcodes and doublets, which
could skew clonotype assignments.</p>
</blockquote>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VDJigsaw_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assign_clonotype.html">assign_clonotype</a></span><span class="op">(</span></span>
<span>  VDJ_data <span class="op">=</span> <span class="va">VDJ.contigs</span>,</span>
<span>  sample_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>  clone_loose <span class="op">=</span> <span class="st">"single_chain_single_allele"</span>,</span>
<span>  num_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [VDJigsaw] ========================================</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Starting VDJigsaw clonotype assignment</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] ========================================</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Detected raw contig input — pivoting to wide format</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Converting VDJ data to wide format...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   Using cdr3_nt for CDR3 chain identifier</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   Pivoting to wide format...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   Wide format: 13639 unique barcodes</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   Creating TRA/TRB chain columns...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Wide format conversion complete: 98 columns</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Validating and correcting TCR data...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Removed 1 duplicate alleles (allele 1 == allele 2)</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Swapped 8 alleles where allele_2 was defined but allele_1 was not</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] TCR data correction complete</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Identifying clonotypes and building reference tables...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Checking inputs...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Processing 13639 cells across 2 sample(s)</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Using 1 core(s) for parallel processing</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Combining results from parallel processing...</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   dual_chain_dual_allele: 7432 clonotypes identified</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   dual_chain_one_partial: 7139 clonotypes identified</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   dual_chain_both_partial: 7117 clonotypes identified</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   single_chain_dual_allele: 7116 clonotypes identified</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw]   single_chain_single_allele: 7116 clonotypes identified</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Clonotype identification complete!</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] ========================================</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Pipeline complete!</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] Assigned 13639 of 13639 cells (100%) to clonotypes</span></span></code></pre>
<pre><code><span><span class="co">## [VDJigsaw] ========================================</span></span></code></pre>
<p>The result is a list with two elements:</p>
<ul>
<li>
<strong><code>TCR_data</code></strong> — wide-format data frame (one
row per cell) with TCR chains and clone IDs at every stringency
level</li>
<li>
<strong><code>ref_tables</code></strong> — named list of reference
tables mapping each <code>CloneID</code> to its chain composition</li>
</ul>
</div>
<div class="section level2">
<h2 id="explore-the-outputs">Explore the outputs<a class="anchor" aria-label="anchor" href="#explore-the-outputs"></a>
</h2>
<div class="section level3">
<h3 id="tcr-data">TCR data<a class="anchor" aria-label="anchor" href="#tcr-data"></a>
</h3>
<p>The <code>TCR_data</code> table has one row per cell barcode, with
clone IDs at every stringency level:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TCR_data</span> <span class="op">&lt;-</span> <span class="va">VDJigsaw_res</span><span class="op">$</span><span class="va">TCR_data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">TCR_data</span><span class="op">)</span>, <span class="st">"cells,"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">TCR_data</span><span class="op">)</span>, <span class="st">"columns\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 13639 cells, 109 columns</span></span></code></pre>
<p>Here are the clone ID columns — each cell gets an assignment at every
stringency level:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clone_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">TCR_data</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^CloneID\\."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">TCR_data</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">TCR_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="st">"barcode"</span>, <span class="va">clone_cols</span><span class="op">)</span><span class="op">]</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="3%">
<col width="9%">
<col width="15%">
<col width="6%">
<col width="15%">
<col width="15%">
<col width="16%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th align="left">Sample</th>
<th align="left">barcode</th>
<th align="left">CloneID.dual_chain_dual_allele</th>
<th align="left">CloneID.loose</th>
<th align="left">CloneID.dual_chain_one_partial</th>
<th align="left">CloneID.dual_chain_both_partial</th>
<th align="left">CloneID.single_chain_dual_allele</th>
<th align="left">CloneID.single_chain_single_allele</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Fresh</td>
<td align="left">AAACCAAAGAACAGAC-1</td>
<td align="left">Clone_Fresh_1</td>
<td align="left">Clone_Fresh_1</td>
<td align="left">Clone_Fresh_1</td>
<td align="left">Clone_Fresh_1</td>
<td align="left">Clone_Fresh_1</td>
<td align="left">Clone_Fresh_1</td>
</tr>
<tr class="even">
<td align="left">Fresh</td>
<td align="left">AAACCAAAGCAAGATA-1</td>
<td align="left">Clone_Fresh_2</td>
<td align="left">Clone_Fresh_2</td>
<td align="left">Clone_Fresh_2</td>
<td align="left">Clone_Fresh_2</td>
<td align="left">Clone_Fresh_2</td>
<td align="left">Clone_Fresh_2</td>
</tr>
<tr class="odd">
<td align="left">Fresh</td>
<td align="left">AAACCAAAGCTGGTTA-1</td>
<td align="left">Clone_Fresh_3</td>
<td align="left">Clone_Fresh_3</td>
<td align="left">Clone_Fresh_3</td>
<td align="left">Clone_Fresh_3</td>
<td align="left">Clone_Fresh_3</td>
<td align="left">Clone_Fresh_3</td>
</tr>
<tr class="even">
<td align="left">Fresh</td>
<td align="left">AAACCAGCACCTAACG-1</td>
<td align="left">Clone_Fresh_4</td>
<td align="left">Clone_Fresh_4</td>
<td align="left">Clone_Fresh_4</td>
<td align="left">Clone_Fresh_4</td>
<td align="left">Clone_Fresh_4</td>
<td align="left">Clone_Fresh_4</td>
</tr>
<tr class="odd">
<td align="left">Fresh</td>
<td align="left">AAACCAGCACGCGTTA-1</td>
<td align="left">Clone_Fresh_5</td>
<td align="left">Clone_Fresh_5</td>
<td align="left">Clone_Fresh_5</td>
<td align="left">Clone_Fresh_5</td>
<td align="left">Clone_Fresh_5</td>
<td align="left">Clone_Fresh_5</td>
</tr>
<tr class="even">
<td align="left">Fresh</td>
<td align="left">AAACCAGCACTAGGAA-1</td>
<td align="left">Clone_Fresh_6</td>
<td align="left">Clone_Fresh_6</td>
<td align="left">Clone_Fresh_6</td>
<td align="left">Clone_Fresh_6</td>
<td align="left">Clone_Fresh_6</td>
<td align="left">Clone_Fresh_6</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="reference-tables">Reference tables<a class="anchor" aria-label="anchor" href="#reference-tables"></a>
</h3>
<p>The <code>ref_tables</code> list contains one reference table per
stringency level. Each table maps a <code>CloneID</code> to its TCR
chain composition.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VDJigsaw_ref_tables</span> <span class="op">&lt;-</span> <span class="va">VDJigsaw_res</span><span class="op">$</span><span class="va">ref_tables</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "dual_chain_dual_allele"     "dual_chain_one_partial"    </span></span>
<span><span class="co">## [3] "dual_chain_both_partial"    "single_chain_dual_allele"  </span></span>
<span><span class="co">## [5] "single_chain_single_allele" "loose"</span></span></code></pre>
<p>Each reference table has columns for <code>CloneID</code>,
<code>Sample</code>, and the TCR chains. Let’s define a small helper to
display them compactly (showing only the V gene part of each chain to
save space):</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Helper: show ref table with truncated chain names for readability</span></span>
<span><span class="va">show_ref</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ref_tbl</span>, <span class="va">n</span> <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">display_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CloneID"</span>, <span class="st">"Sample"</span>, <span class="st">"TRA_1"</span>, <span class="st">"TRA_2"</span>, <span class="st">"TRB_1"</span>, <span class="st">"TRB_2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ref_tbl</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ref_tbl</span><span class="op">[</span>, <span class="va">display_cols</span><span class="op">]</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="co"># Shorten V_CDR3_J strings to just V gene for display</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">col</span> <span class="kw">in</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRA_1"</span>, <span class="st">"TRA_2"</span>, <span class="st">"TRB_1"</span>, <span class="st">"TRB_2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">df</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_.*"</span>, <span class="st">""</span>, <span class="va">df</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">df</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level4">
<h4 class="tabset" id="stringency-levels">Stringency levels<a class="anchor" aria-label="anchor" href="#stringency-levels"></a>
</h4>
<p>The stringency levels are named along two axes: <strong>which chains
are required</strong> (both alpha and beta, or just one) and <strong>how
many alleles per chain</strong> (both alleles, or just one).</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="12%">
<col width="33%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th>Level</th>
<th>Name</th>
<th>Chains required</th>
<th>Alleles per chain</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td><code>dual_chain_dual_allele</code></td>
<td>Both alpha and beta</td>
<td>Both alleles (1 and 2)</td>
</tr>
<tr class="even">
<td>2</td>
<td><code>dual_chain_one_partial</code></td>
<td>Both alpha and beta</td>
<td>One chain can have a single allele</td>
</tr>
<tr class="odd">
<td>3</td>
<td><code>dual_chain_both_partial</code></td>
<td>Both alpha and beta</td>
<td>Both chains can have a single allele</td>
</tr>
<tr class="even">
<td>4</td>
<td><code>single_chain_dual_allele</code></td>
<td>alpha or beta alone</td>
<td>Both alleles (1 and 2)</td>
</tr>
<tr class="odd">
<td>5</td>
<td><code>single_chain_single_allele</code></td>
<td>alpha or beta alone</td>
<td>A single allele suffices</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="dual_chain_dual_allele">dual_chain_dual_allele<a class="anchor" aria-label="anchor" href="#dual_chain_dual_allele"></a>
</h5>
<p>The strictest level. All four chain positions (<code>TRA_1</code>,
<code>TRA_2</code>, <code>TRB_1</code>, <code>TRB_2</code>) must match
exactly.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_dual_allele</span><span class="op">)</span>, <span class="st">"clonotypes\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 7432 clonotypes</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">show_ref</span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_dual_allele</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CloneID</th>
<th align="left">Sample</th>
<th align="left">TRA_1</th>
<th align="left">TRA_2</th>
<th align="left">TRB_1</th>
<th align="left">TRB_2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Clone_Fresh_1</td>
<td align="left">Fresh</td>
<td align="left">TRAV19</td>
<td align="left">TRAV8-4</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_10</td>
<td align="left">Fresh</td>
<td align="left">TRAV6</td>
<td align="left">NA</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_100</td>
<td align="left">Fresh</td>
<td align="left">TRAV12-1</td>
<td align="left">TRAV21</td>
<td align="left">TRBV19</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_1000</td>
<td align="left">Fresh</td>
<td align="left">TRAV8-6</td>
<td align="left">NA</td>
<td align="left">TRBV28</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_1001</td>
<td align="left">Fresh</td>
<td align="left">TRAV8-6</td>
<td align="left">NA</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="dual_chain_one_partial">dual_chain_one_partial<a class="anchor" aria-label="anchor" href="#dual_chain_one_partial"></a>
</h5>
<p>Both chains required, but one chain may have a missing secondary
allele (no conflict allowed).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_one_partial</span><span class="op">)</span>, <span class="st">"clonotypes\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 7139 clonotypes</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">show_ref</span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_one_partial</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CloneID</th>
<th align="left">Sample</th>
<th align="left">TRA_1</th>
<th align="left">TRA_2</th>
<th align="left">TRB_1</th>
<th align="left">TRB_2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Clone_Fresh_1</td>
<td align="left">Fresh</td>
<td align="left">TRAV19</td>
<td align="left">TRAV8-4</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_10</td>
<td align="left">Fresh</td>
<td align="left">TRAV6</td>
<td align="left">NA</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_100</td>
<td align="left">Fresh</td>
<td align="left">TRAV1-2</td>
<td align="left">NA</td>
<td align="left">TRBV6-5</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_1000</td>
<td align="left">Fresh</td>
<td align="left">TRAV19</td>
<td align="left">NA</td>
<td align="left">TRBV6-5</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_1001</td>
<td align="left">Fresh</td>
<td align="left">TRAV1-2</td>
<td align="left">NA</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="dual_chain_both_partial">dual_chain_both_partial<a class="anchor" aria-label="anchor" href="#dual_chain_both_partial"></a>
</h5>
<p>Both chains required, but both may have a missing secondary
allele.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_both_partial</span><span class="op">)</span>, <span class="st">"clonotypes\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 7117 clonotypes</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">show_ref</span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_both_partial</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CloneID</th>
<th align="left">Sample</th>
<th align="left">TRA_1</th>
<th align="left">TRA_2</th>
<th align="left">TRB_1</th>
<th align="left">TRB_2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Clone_Fresh_1</td>
<td align="left">Fresh</td>
<td align="left">TRAV19</td>
<td align="left">TRAV8-4</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_10</td>
<td align="left">Fresh</td>
<td align="left">TRAV6</td>
<td align="left">NA</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_100</td>
<td align="left">Fresh</td>
<td align="left">TRAV1-2</td>
<td align="left">NA</td>
<td align="left">TRBV6-5</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_1000</td>
<td align="left">Fresh</td>
<td align="left">TRAV1-2</td>
<td align="left">TRAV8-3</td>
<td align="left">TRBV12-4</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_1001</td>
<td align="left">Fresh</td>
<td align="left">TRAV8-1</td>
<td align="left">NA</td>
<td align="left">TRBV5-1</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="single_chain_dual_allele">single_chain_dual_allele<a class="anchor" aria-label="anchor" href="#single_chain_dual_allele"></a>
</h5>
<p>Only one chain (alpha or beta) is required, but both alleles of that
chain must match.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">single_chain_dual_allele</span><span class="op">)</span>, <span class="st">"clonotypes\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 7116 clonotypes</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">show_ref</span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">single_chain_dual_allele</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CloneID</th>
<th align="left">Sample</th>
<th align="left">TRA_1</th>
<th align="left">TRA_2</th>
<th align="left">TRB_1</th>
<th align="left">TRB_2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Clone_Fresh_1</td>
<td align="left">Fresh</td>
<td align="left">TRAV19</td>
<td align="left">TRAV8-4</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_10</td>
<td align="left">Fresh</td>
<td align="left">TRAV6</td>
<td align="left">NA</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_100</td>
<td align="left">Fresh</td>
<td align="left">TRAV1-2</td>
<td align="left">NA</td>
<td align="left">TRBV6-5</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_1000</td>
<td align="left">Fresh</td>
<td align="left">TRAV1-2</td>
<td align="left">TRAV8-3</td>
<td align="left">TRBV12-4</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_1001</td>
<td align="left">Fresh</td>
<td align="left">TRAV8-1</td>
<td align="left">NA</td>
<td align="left">TRBV5-1</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level5">
<h5 id="single_chain_single_allele">single_chain_single_allele<a class="anchor" aria-label="anchor" href="#single_chain_single_allele"></a>
</h5>
<p>The most permissive level. A single allele of a single chain can
define a clonotype. Maximizes cell recovery but should be used with
caution.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">single_chain_single_allele</span><span class="op">)</span>, <span class="st">"clonotypes\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 7116 clonotypes</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">show_ref</span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">single_chain_single_allele</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CloneID</th>
<th align="left">Sample</th>
<th align="left">TRA_1</th>
<th align="left">TRA_2</th>
<th align="left">TRB_1</th>
<th align="left">TRB_2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Clone_Fresh_1</td>
<td align="left">Fresh</td>
<td align="left">TRAV19</td>
<td align="left">TRAV8-4</td>
<td align="left">TRBV20-1</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_10</td>
<td align="left">Fresh</td>
<td align="left">TRAV2</td>
<td align="left">NA</td>
<td align="left">TRBV14</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_100</td>
<td align="left">Fresh</td>
<td align="left">TRAV26-1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">Clone_Fresh_1000</td>
<td align="left">Fresh</td>
<td align="left">TRAV21</td>
<td align="left">NA</td>
<td align="left">TRBV6-5</td>
<td align="left">NA</td>
</tr>
<tr class="odd">
<td align="left">Clone_Fresh_1001</td>
<td align="left">Fresh</td>
<td align="left">TRAV26-2</td>
<td align="left">NA</td>
<td align="left">TRBV5-4</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="section"></h4>
<p>Let’s see how the number of clonotypes changes across levels:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">VDJigsaw_ref_tables</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="va">nrow</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     dual_chain_dual_allele     dual_chain_one_partial </span></span>
<span><span class="co">##                       7432                       7139 </span></span>
<span><span class="co">##    dual_chain_both_partial   single_chain_dual_allele </span></span>
<span><span class="co">##                       7117                       7116 </span></span>
<span><span class="co">## single_chain_single_allele </span></span>
<span><span class="co">##                       7116</span></span></code></pre>
<p>As expected, the number of unique clonotypes decreases as stringency
is relaxed (cells get merged into larger groups).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualize-results">Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h2>
<p>VDJigsaw includes three built-in visualization functions. All support
<code>per_sample</code> and <code>combined</code> arguments for
multi-sample datasets.</p>
<div class="section level3">
<h3 id="stringency-summary">Stringency summary<a class="anchor" aria-label="anchor" href="#stringency-summary"></a>
</h3>
<p>This two-panel chart shows (top) the number of unique clonotypes and
(bottom) the number of cells assigned at each stringency level.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_stringency_summary.html">plot_stringency_summary</a></span><span class="op">(</span></span>
<span>  TCR_data <span class="op">=</span> <span class="va">TCR_data</span>,</span>
<span>  per_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  combined <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="VDJigsaw_Vignette_01_files/figure-html/unnamed-chunk-18-1.png" class="r-plt" alt="" width="680"></p>
</div>
<div class="section level3">
<h3 id="clonotype-composition">Clonotype composition<a class="anchor" aria-label="anchor" href="#clonotype-composition"></a>
</h3>
<p>A tile chart showing V gene usage for the top clonotypes at a given
stringency level. Grey tiles indicate missing chains.</p>
<div class="section level4">
<h4 id="strict-definition">Strict definition<a class="anchor" aria-label="anchor" href="#strict-definition"></a>
</h4>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_clone_composition.html">plot_clone_composition</a></span><span class="op">(</span></span>
<span>  ref_table <span class="op">=</span> <span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">dual_chain_dual_allele</span>,</span>
<span>  top_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  per_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  combined <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_list</span><span class="op">$</span><span class="va">Fix</span> <span class="op">+</span></span>
<span>  <span class="va">plot_list</span><span class="op">$</span><span class="va">Fresh</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="VDJigsaw_Vignette_01_files/figure-html/unnamed-chunk-19-1.png" class="r-plt" alt="" width="680"></p>
</div>
<div class="section level4">
<h4 id="loose-definition">Loose definition<a class="anchor" aria-label="anchor" href="#loose-definition"></a>
</h4>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_clone_composition.html">plot_clone_composition</a></span><span class="op">(</span></span>
<span>  ref_table <span class="op">=</span> <span class="va">VDJigsaw_ref_tables</span><span class="op">$</span><span class="va">single_chain_single_allele</span>,</span>
<span>  top_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  per_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  combined <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_list</span><span class="op">$</span><span class="va">Fix</span> <span class="op">+</span></span>
<span>  <span class="va">plot_list</span><span class="op">$</span><span class="va">Fresh</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="VDJigsaw_Vignette_01_files/figure-html/unnamed-chunk-20-1.png" class="r-plt" alt="" width="680"></p>
</div>
</div>
<div class="section level3">
<h3 id="clonotype-flow-alluvial">Clonotype flow (alluvial)<a class="anchor" aria-label="anchor" href="#clonotype-flow-alluvial"></a>
</h3>
<p>An alluvial diagram showing how clonotypes merge as stringency is
relaxed. Each vertical axis is a stringency level, and flows show cells
moving between clone definitions.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_clonotype_flow.html">plot_clonotype_flow</a></span><span class="op">(</span></span>
<span>  TCR_data <span class="op">=</span> <span class="va">TCR_data</span>,</span>
<span>  top_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  per_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  combined <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_list</span><span class="op">$</span><span class="va">Fix</span> <span class="op">+</span></span>
<span>  <span class="va">plot_list</span><span class="op">$</span><span class="va">Fresh</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="VDJigsaw_Vignette_01_files/figure-html/unnamed-chunk-21-1.png" class="r-plt" alt="" width="680"></p>
</div>
</div>
<div class="section level2">
<h2 id="next-steps">Next steps<a class="anchor" aria-label="anchor" href="#next-steps"></a>
</h2>
<ul>
<li>
<strong>Integrate with Seurat:</strong> Map <code>TCR_data</code>
back to your Seurat object using the <code>barcode</code> column to
annotate cells with clone IDs.</li>
<li>
<strong>Cross-reference:</strong> Use
<code><a href="../reference/assign_clonotype_from_reference.html">assign_clonotype_from_reference()</a></code> to map new data against
an existing reference clonotype table.</li>
<li>
<strong>Paired TCR mapping:</strong> Use
<code><a href="../reference/map_clonotypes_to_paired_TCR.html">map_clonotypes_to_paired_TCR()</a></code> to match allele-level data
to a paired alpha/beta reference.</li>
</ul>
<p>For more details, see the function documentation:
<code><a href="../reference/assign_clonotype.html">?assign_clonotype</a></code>,
<code><a href="../reference/assign_clonotype_from_reference.html">?assign_clonotype_from_reference</a></code>,
<code><a href="../reference/map_clonotypes_to_paired_TCR.html">?map_clonotypes_to_paired_TCR</a></code>.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] patchwork_1.3.2 ggplot2_4.0.2   dplyr_1.2.0     VDJigsaw_0.1.0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.10        generics_0.1.4     tidyr_1.3.2        stringi_1.8.7     </span></span>
<span><span class="co">##  [5] digest_0.6.39      magrittr_2.0.4     evaluate_1.0.5     grid_4.5.2        </span></span>
<span><span class="co">##  [9] RColorBrewer_1.1-3 iterators_1.0.14   fastmap_1.2.0      foreach_1.5.2     </span></span>
<span><span class="co">## [13] doParallel_1.0.17  jsonlite_2.0.0     purrr_1.2.1        scales_1.4.0      </span></span>
<span><span class="co">## [17] codetools_0.2-20   textshaping_1.0.4  jquerylib_0.1.4    cli_3.6.5         </span></span>
<span><span class="co">## [21] rlang_1.1.7        withr_3.0.2        cachem_1.1.0       yaml_2.3.12       </span></span>
<span><span class="co">## [25] tools_4.5.2        parallel_4.5.2     vctrs_0.7.1        R6_2.6.1          </span></span>
<span><span class="co">## [29] lifecycle_1.0.5    stringr_1.6.0      fs_1.6.6           ragg_1.5.0        </span></span>
<span><span class="co">## [33] pkgconfig_2.0.3    desc_1.4.3         pkgdown_2.2.0      pillar_1.11.1     </span></span>
<span><span class="co">## [37] bslib_0.10.0       gtable_0.3.6       glue_1.8.0         systemfonts_1.3.1 </span></span>
<span><span class="co">## [41] xfun_0.56          tibble_3.3.1       tidyselect_1.2.1   knitr_1.51        </span></span>
<span><span class="co">## [45] farver_2.1.2       htmltools_0.5.9    igraph_2.2.2       rmarkdown_2.30    </span></span>
<span><span class="co">## [49] labeling_0.4.3     ggalluvial_0.12.5  compiler_4.5.2     S7_0.2.1</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rémy Pétremand.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
